moonraker/components/prometheus.py

```diff
@@ -0,0 +1,157 @@
+        kapi: KlippyAPI = self.server.lookup_component("klippy_apis")
+        result = await kapi.query_objects(objs)
+        await self._handle_status_update(result, None)
+
+    async def _handle_klippy_started(self, state: KlippyState) -> None:
+        """Upon klippy startup, it queries current statuses
+        and subscribes for updates."""
+        self._clear_metrics()
+
+        kapi: KlippyAPI = self.server.lookup_component("klippy_apis")
+        subs = await self._get_objects_to_subscribe()
+
+        await gather(
+            self._init_metrics(subs),
+            kapi.subscribe_objects(subs, self._handle_status_update)
+        )
+
+        logging.info("Prometheus handler registered and subscribed to status updates")
+
+    async def _handle_status_update(self, status: Dict[str, Dict[str, Any]],
+                                    eventtime: float | None) -> None:
+        for key, value in status.items():
+            module = key.split()[0]
+            if module in ['heater_bed', 'extruder', 'heater_generic']:
+                self._status_update_heater(key, value)
+            elif module in ['temperature_combined', 'temperature_sensor', 'tmc2240',
+                            'temperature_fan']:
+                self._status_update_temp_sensor(key, value)
+            else:
+                logging.debug("[prometheus]: unhandled status for object %s" % key)
+
+    def _status_update_temp_sensor(self, sensor_name: str,
+                                   status: Dict[str, Any]) -> None:
+        temp = status.get('temperature', None)
+        if temp is not None:
+            self.m_temp.labels(sensor_name).set(temp)
+
+    def _status_update_heater(self, heater_name: str,
+                              status: Dict[str, Any]) -> None:
+        temp = status.get('temperature', None)
+        if temp is not None:
+            self.m_temp.labels(heater_name).set(temp)
+
+        target = status.get('target', None)
+        if target is not None:
+            self.m_target_temp.labels(heater_name).set(target)
+
+        power = status.get('power', None)
+        if power is not None:
+            self.m_heater_power.labels(heater_name).set(power)
+
+    def _clear_metrics(self) -> None:
+        for attr_name, attr_value in self.__dict__.items():
+            if isinstance(attr_value, MetricWrapperBase):
+                attr_value.clear()
+
+    async def _handle_metrics_endpoint(self, web_request: WebRequest) -> bytes:
+        """Writes metrics in response to the scrape.
+
+        Usually some properties of the response depend on request headers.
+        To make request headers available here, a serious refactoring would be needed.
+        Instead, we make some assumptions:
+        - Response will be in the "normal" format and not openmetrics
+        - No filtering by name will be done
+        - gzip won't be used (also sparing the CPU cycles in return for bandwidth)
```

Comment:
It wouldn't be difficult to provide the `WebRequest` access to the request headers if needed.  This hasn't been required to date so I haven't done so.

As far as gzip compression, a reverse proxy could handle that if its desired.  I don't think it needs to be in Moonraker.

Reply:
> As far as gzip compression, a reverse proxy could handle that if its desired. I don't think it needs to be in Moonraker.
You are absolutely right. These notes basically address all the parts of prometheus client's default handler that I stripped in order to get this integrated.